File: python/meetup/meetup.py
from datetime import date

class MeetupDayException(Exception):
  pass

DAYS_OF_WEEK = {
  'Sunday': 0,
  'Monday': 1,
  'Tuesday': 2,
  'Wednesday': 3,
  'Thursday': 4,
  'Friday': 5,
  'Saturday': 6,
}

def meetup_day(year, month, day_of_the_week, which):
  month_days = [
    None, 31,
    29 if month == 2 and year % 4 == 0 and (year % 100 != 0 or year % 400 == 0) else 28,
    31, 30, 31, 30, 31, 31, 30, 31, 30, 31
  ]

  # get day of week of the 1st of the month
  first_day_of_month = int(date(year, month, 1).strftime("%w"))

  # get first instance of target day in month
  first_target_day_num = DAYS_OF_WEEK[day_of_the_week]

  target_date = 1 + first_target_day_num - first_day_of_month \
    if first_target_day_num >= first_day_of_month \
    else 1 + (7 - first_day_of_month) + first_target_day_num

  if which == 'teenth':
    target_date += 7 if target_date > 5 else 14
  elif which == '2nd':
    target_date += 7
  elif which == '3rd':
    target_date += 14
  elif which == '4th':
    target_date += 21
  elif which == 'last':
    target_date += 21 if target_date + 28 > month_days[month] else 28
  elif which == '5th':
    if target_date + 28 > month_days[month]:
        raise MeetupDayException('Invalid 5th day')
    target_date += 28
  
  return date(year, month, target_date)

File: csharp/meetup/Meetup.cs
using System;

public enum Schedule
{
    Teenth,
    First,
    Second,
    Third,
    Fourth,
    Last
}

public class Meetup
{
    readonly int month;
    readonly int year;
    public Meetup(int month, int year)
    {
        this.month = month;
        this.year = year;
    }

    public DateTime Day(DayOfWeek dayOfWeek, Schedule schedule)
    {
        DayOfWeek firstDayOfMonth = new DateTime(year, month, 1).DayOfWeek;

        // check if target day is before or after first day of month
        int dateOfFirstTargetDayOfWeek = dayOfWeek >= firstDayOfMonth ? (
            1 + (int)dayOfWeek - (int)firstDayOfMonth
        ) : (
            7 - (int)firstDayOfMonth + 1 + (int)dayOfWeek
        );
        int day = dateOfFirstTargetDayOfWeek;
        switch (schedule) {
            case Schedule.Teenth:
                day = day < 6 ? day + 14 : day + 7;
                break;
            case Schedule.Second:
                day += 7;
                break;
            case Schedule.Third:
                day += 14;
                break;
            case Schedule.Fourth:
                day += 21;
                break;
            case Schedule.Last:
                if (month == 2) {
                    // leap year
                    if (year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)) {
                        day += day == 1 ? 28 : 21;
                    } else {
                        day += 21;
                    }
                } else {
                    if (DateTime.DaysInMonth(year, month) == 30) {
                        day += day < 3 ? 28 : 21;
                    } else {
                        day += day < 4 ? 28 : 21;
                    }
                }
                break;
            default:
                break;
        }

        return new DateTime(year, month, day);
    }
}